Fix parsing nullable collections

hemberger opened the pull request:
Fixes #163.
Fixes phpDocumentor/phpDocumentor#3290.

Shorthand nullable type syntax is now supported for collection types
(e.g. `?array<int>`).

This also prevents a misuse of the shorthand nullable type syntax with
compound types, which is illegal with PHP types. This is documented in
the Union Types 2.0 RFC (https://wiki.php.net/rfc/union_types_v2) by
N. Popov:

> Union types and the ?T nullable type notation cannot be mixed.
> Writing ?T1|T2, T1|?T2 or ?(T1|T2) is not supported and T1|T2|null
> needs to be used instead.

hemberger committed:
Fix parsing nullable collections

Fixes #163.
Fixes phpDocumentor/phpDocumentor#3290.

Shorthand nullable type syntax is now supported for collection types
(e.g. `?array<int>`).

This also prevents a misuse of the shorthand nullable type syntax with
compound types, which is illegal with PHP types. This is documented in
the Union Types 2.0 RFC (https://wiki.php.net/rfc/union_types_v2) by
N. Popov:

> Union types and the ?T nullable type notation cannot be mixed.
> Writing ?T1|T2, T1|?T2 or ?(T1|T2) is not supported and T1|T2|null
> needs to be used instead.

jaapio commented on tests/unit/TypeResolverTest.php:
```diff
@@ -407,9 +407,9 @@ public function testResolvingNullableCompoundTypes(): void
+        $this->expectExceptionMessage('Unexpected type separator');
         $resolvedType = $fixture->resolve('?string|null|?boolean');
-
-        $this->assertSame('?string|null|?bool', (string) $resolvedType);
```
I'm not sure if this behavior should change. I do understand that PHP doesn't allow this format and that it looks more than ugly. But this case has been added because we found some repositories using this notation. 

As this library is used to parse types in docblocks and native PHP, I think we should keep this format supported. 
It could result in a deprecation that we will remove in the future. But for now, this is a backward compatibility break which we cannot do with the number of consuming projects for this library. 

Do you think it is possible to revert this change?

hemberger replied:
Thanks for the feedback! I did not fully appreciate the need for backwards compatibility, so I have restored support for shorthand nullable syntax in compound types as requested. Please let me know if this solution is not satisfactory.

hemberger committed:
Restore support for shorthand nullable compounds

While PHP types do not support shorthand nullable syntax in compound
types (e.g. `?string|int`), it has been requested to retain support
for this in phpdoc types for backwards compatibility reasons.
